Add webhook triggered on Order Cycle Open

dacook opened the pull request:
### What? Why?
- Closes #9616
   > As an API user I want to be notified when an OC opens.
    
A webhook can be sent to a service like Zapier, which can generate an email notification.

### How?
Currently, we have no way of sending a webhook. There's no pre-built way of doing this for Rails surprisingly, although [this tutorial](https://keygen.sh/blog/how-to-build-a-webhook-system-in-rails-using-sidekiq/#handing-delivery-timeouts) looks like a good formula. I followed this, but found some parts unnecessary, and ended up with a different solution.

My focus was on making this work firstly for this single event type. Of course we'd like to extend it to for other events in the future, but let's get one working first before making further decisions.

There's bit going on here. A quick summary:
1. **WebhookDeliveryJob**: a job that will be enqueued to deliver a payload for each triggered webhook
2. **WebhookEndpoint**: a model to store each webhook endpoint url (in the future it would also store the event type it is listening for).
3. **OrderCycleWebhookService**: create webhook payloads for an order cycle event.
4. **OrderCycleOpenedJob**: Uses OrderCycleWebhookService to schedule WebhookDeliveryJobs. The order cycle remembers if it's been opened, and can only open once. 
5. (removed)
6. **User account interface to manage webhooks**. I was surprised how hard it was to fiddle around and keep the form simple, but got there in the end. 
    - Rather than enable/disable as originally suggested, you can just create or delete which has the same effect.
    - a test mode hasn't been implemented  yet, in order to keep this PR from dragging out. 

|||
|---|---|
|<a target="_blank" rel="noopener noreferrer nofollow" href="https://user-images.githubusercontent.com/4188088/223575770-879ca807-e87d-428c-b5e5-36895c7d7439.png"><img src="https://user-images.githubusercontent.com/4188088/223575770-879ca807-e87d-428c-b5e5-36895c7d7439.png" alt="Screen Shot 2023-03-08 at 10 14 45 am" width="400" style="max-width: 100%;"></a>|<a target="_blank" rel="noopener noreferrer nofollow" href="https://user-images.githubusercontent.com/4188088/223576129-8a56fc13-9780-44f0-b8b9-ee588eff862e.png"><img src="https://user-images.githubusercontent.com/4188088/223576129-8a56fc13-9780-44f0-b8b9-ee588eff862e.png" alt="Screen Shot 2023-03-08 at 10 15 58 am" width="400" style="max-width: 100%;"></a><a target="_blank" rel="noopener noreferrer nofollow" href="https://user-images.githubusercontent.com/4188088/223576140-ef99a59c-7ac5-477b-b67d-e1819cbf09b9.png"><img src="https://user-images.githubusercontent.com/4188088/223576140-ef99a59c-7ac5-477b-b67d-e1819cbf09b9.png" alt="Screen Shot 2023-03-08 at 10 16 27 am" width="400" style="max-width: 100%;"></a>|

### What should we test?
This is a new feature which doesn't touch any existing functionality.
It adds a new form to the Developer tab on the user Accounts page. 
Suggested tests:

1. **Prepare a webhook URL** ready to receive and capture a webhook payload. The easiest way is:
    - Visit https://webhook.site/
    - Copy **Your unique URL**
    - Keep the tab open to check results later.
2. **Create/edit a user with API access:** 
    - Admin > Users > Edit
    - Tick "SHOW API KEY VIEW FOR USER", save.
3. **Login as this user**
6. **Go to My account > Developer settings > Webhook Endpoints**
    - [x] Verify the interface as per [requirements](https://github.com/openfoodfoundation/openfoodnetwork/issues/9616#issue-1357577999).
8. **Enter your unique Webhook URL and click create**
    - [x] Verify that the URL has been saved.
9. **Set up an Order Cycle to open at the current time** (anytime in the last hour):
    - Admin > Order Cycles > + New Order Cycle
    - Choose co-ordinator and click continue
    - Enter a name, and choose a time that is about to pass
    - Save (no need to complete further steps).
10. **Wait** up to 5 minutes for the webhook event to be captured ☕️ 
    (refer back to the open tab for https://webhook.site/)
    - [x] Verify the result as per [requirements](https://github.com/openfoodfoundation/openfoodnetwork/issues/9616#issue-1357577999)
12. **De-activate the webhook:**
    - Go to My account > Developer settings > Webhook Endpoints
    - Delete the webhook endpoint URL.
9. **Set up another Order Cycle to open at the current time**
13. **Wait** minimum 5 mins
    - [x] Verify that no event was received.


### Release notes

Changelog Category: User facing changes

_If documentation has been written, share a link in the release notes_

<!-- #### Dependencies -->

### Documentation updates
I think it should be documented wherever the API is documented. But I can't find any public mention of the API??!
It seems worth mentioning in the user guide: https://guide.openfoodnetwork.org/

Ideally, a test mode will display the webhook payload format, and act as in-place documentation. But this will come in a future PR.

In the future, [I suggest we integrate with our OpenAPI documentation too](https://github.com/openfoodfoundation/openfoodnetwork/issues/9616#issuecomment-1425133692).

dacook commented:
HI @jibees or @mkllnk, this is not yet finished, but it would be great to have someone's eyes over it to make sure we're heading in the right direction. 
Would you mind starting perhaps by reviewing the commit messages and let me know if you have any concerns, or you're happy for me to proceed? (the first six commits are from another PR and can be ignored)

You can see the further introduction and screenshot in description above.

mkllnk commented on app/models/enterprise.rb:
```diff
@@ -64,6 +64,7 @@ class Enterprise < ApplicationRecord
   has_many :customers
   has_many :inventory_items
   has_many :tag_rules
+  belongs_to :webhook_endpoint, optional: true, dependent: :destroy
```
Would it be better the other way round? One endpoint belongs to an enterprise and an enterprise *can* have one endpoint. It avoids an optional, sparse column on enterprise because the endpoint definitely requires an owner. And we don't want enterprises to share endpoints. It also means that the schema doesn't need changing if we want to change from has_one to has_many.

dacook replied:
Much better, yes! Updated.

mkllnk commented on app/jobs/order_cycle_webhook_job.rb:
```diff
@@ -0,0 +1,21 @@
+# Creates a webhook payload for an order cycle
+class OrderCycleWebhookJob < ApplicationJob
+  def perform(order_cycle_id, event)
+    order_cycle = OrderCycle.find(order_cycle_id)
```
If the OrderCycle has been deleted, this fails and will be retried. We have that with some other job already. So maybe rescue, log and return without error.

dacook replied:
Good idea, done.

mkllnk commented on app/jobs/order_cycle_webhook_job.rb:
```diff
@@ -0,0 +1,21 @@
+  def perform(order_cycle_id, event)
+    order_cycle = OrderCycle.find(order_cycle_id)
+    data = order_cycle.slice(:id, :name).merge(
+      {
+        orders_open_at: order_cycle.orders_open_at&.to_s,
+        orders_close_at: order_cycle.orders_close_at&.to_s,
+        enterprise_id: order_cycle.coordinator_id,
+        enterprise_name: order_cycle.coordinator.name,
+      }
```
Why this translation? Could you just slice all of them? And why to_s?

dacook replied:
> why to_s?

I can't remember why, but I was making OrderCycleWebhookJob responsible for formatting the dates. I don't think that's necessary, it can be handled by WebhookDeliveryJob (with default formatting anyway).

> Could you just slice all of them?

* `orders_open/close_at`: yep, now with `to_s` removed
* `coordinator_id`: yep, I think it would be better to use the underlying name `coordinator`.
* `coordinator.name`: As far as i know there's no builtin way to 'deep slice', so I think this translation is still necessary.

mkllnk replied:
> `coordinator.name`

Oh, of course. I overlooked that one.

mkllnk commented on app/jobs/order_cycle_webhook_job.rb:
```diff
@@ -0,0 +1,21 @@
+      }
+    )
+
+    endpoint = order_cycle.coordinator.webhook_endpoint
```
Interesting. I could imagine that it would be very useful to loop through all distributors in that order cycle. Maybe clarify with product. One odd case is that the coordinator doesn't need to be selling within the OC. So it's like two different events or subscriptions. The enterprise may want to choose to receive events as coordinator, as shop or as producer. So maybe this is just the easiest for now. Clarify with product?

dacook replied:
I didn't think of that, good idea.
Do you mean to clarify in in the #product-circle channel?

To start with, I'll raise the question with Lynne on the issue.

dacook replied:
https://github.com/openfoodfoundation/openfoodnetwork/issues/9616#issuecomment-1271011028

mkllnk commented on spec/jobs/order_cycle_webhook_job_spec.rb:
```diff
@@ -0,0 +1,67 @@
+
+  context "subscribed to different event" do
+    it "doesn't create webhook payload" do
+      expect(WebhookDeliveryJob).not_to receive(:perform_later)
+        .with("http://url", any_args)
```
I guess that this would pass if we accidentally enqueued a job with a different URL? Shall we test for enqueued jobs instead?

mkllnk replied:
Mocking or testing for method calls is always a bit brittle. Better to test for outcomes (job in the database).

dacook replied:
> I guess that this would pass if we accidentally enqueued a job with a different URL? 

Yeah, I think we should test that no jobs were created, regardless of url. Updated.

> Better to test for outcomes (job in the database).

Totally. I was having trouble and gave up before. Don't know why though, it seems obvious now. Done.

mkllnk commented on app/jobs/order_cycle_webhook_job.rb:
```diff
@@ -4,6 +4,11 @@
 class OrderCycleWebhookJob < ApplicationJob
   def perform(order_cycle_id, event)
     order_cycle = OrderCycle.find(order_cycle_id)
+    return if already_notified?(order_cycle, event)
```
This is prone to a race condition. Maybe we do it in one query like `OrderCycle.lock.opened_unnotified.find`?

Reading this I'm getting a bit confused about the different jobs. Maybe they need an explanation in a comment at the top. Who's triggering them when and what do they trigger.

mkllnk commented on app/jobs/order_cycle_opening_job.rb:
```diff
@@ -0,0 +1,18 @@
+# frozen_string_literal: true
+
+# Trigger jobs for any order cycles that recently opened
+class OrderCycleOpeningJob < ApplicationJob
```
`OrderCycleOpenedJob`? Otherwise you could read it as the job that opens the order cycle.

dacook replied:
Yeah that's a good question. I err-ed on the side of following Matt's precedent with `OrderCycleClosingJob`.

But I think you're right: the order cycle is state considered changed based on time, not the actions performed in these jobs. Some actions are performed, but retrospectively _after_ the (implicit) state change. So would you agree we should also rename it to OrderCycleClos***ed***Job?

dacook replied:
Agreed, but to rename the existing job in a separate PR. It will require a transition, eg keep the existing named class as a sub-class for compatibility with jobs still in the queue.

mkllnk commented on app/jobs/order_cycle_opening_job.rb:
```diff
@@ -0,0 +1,18 @@
+    recently_opened_order_cycles.each do |oc_id|
+      SubscriptionPlacementJob.perform_later(oc_id)
+    end
+  end
```
I imagined the timestamp to be set here. It's like the initial event as been dispatched. It can be done in bulk and it prevents more jobs from being enqueued which then cancel themselves because they have been executed already.

dacook replied:
I think I'd like to discuss this again, as I'm not sure about it. 

I find myself wondering if indeed we do need an intermediate job (OrderCycleWebhookJob) simply for creating a tiny data blob (payload)  and scheduling for delivery. 
Yes, perhaps if there were hundreds of order cycles with webhooks opening at the same time, it might take a bit longer. And maybe if having instant notifications was crucial, this could pose a problem. But I don't think we're anywhere near the need for that optimisation. 

So I think I would move that logic into OrderCycleOpeningJob. Let's discuss when we're both available.

mkllnk reviewed (commented):
Great. This all makes sense. I couldn't resist looking at the code though and left some comments below. You may have thought of it all already.

dacook commented on db/migrate/20220916045836_create_webhook_endpoints.rb:
```diff
@@ -0,0 +1,12 @@
+  def change
+    create_table :webhook_endpoints do |t|
+      t.string :url, null: false
+      t.string :subscriptions, null: false
```
Good idea, done. This name was originally based on keygen's model which allowed subscribing to multiple events. But I decided not to support that yet.

mkllnk commented on app/jobs/order_cycle_opening_job.rb:
```diff
@@ -0,0 +1,19 @@
+  private
+
+  def recently_opened_order_cycles
+    OrderCycle
+      .where(orders_open_at: 1.hour.ago..Time.zone.now)
+      .pluck(:id)
```
```suggestion
    OrderCycle
      .where(orders_open_at: 1.hour.ago..Time.zone.now)
      .lock.order(:id).pluck(:id)
```

dacook commented on app/controllers/spree/users_controller.rb:
```diff
@@ -94,7 +100,16 @@ def accurate_title
+      @user_params ||= ::PermittedAttributes::User.new(params).call
+    end
+
+    def strip_blank_webhook_endpoints
+      # Ignore the new record (with no id) if there is no data entered (empty params)
+      user_params[:webhook_endpoints_attributes]&.each do |key, record|
+        if record[:id].blank? && record[:url].blank?
+          user_params[:webhook_endpoints_attributes].delete key
+        end
+      end
```
I borrowed this method from Admin::EnterprisesController#strip_new_properties

dacook commented:
Hi @mkllnk ,  I'd like to validate this approach that I've tried for the form, in this commit: https://github.com/openfoodfoundation/openfoodnetwork/pull/9687/commits/ef82335ebafa49a45caff976431c68f87a623574. As I already had the code ready for multiple endpoints, I've re-used this to add a simple tabular form. I intend to add an action column with delete link, and from there it would be simple to allow multiple endpoints.


There's a couple of caveats but I think they're ok.
* after `UsersController#update` succeeds, it always returns to the first tab (orders) with message "Account updated". But that's already the case for updating account settings.
* if `UsersController#update` fails (eg due to validation error), it renders the edit form, which only shows user account settings.  I think that hardly matters though, given that I've made use of browser validation to ensure the field is non-empty and contains a url.

Edit screen:

![Screen Shot 2022-11-04 at 5 30 01 pm](https://user-images.githubusercontent.com/4188088/199906595-60ee38b8-8bdd-4529-ab01-4bc400bdfc4d.png)

Validation error (although I guarded against this particular error)
![Screen Shot 2022-11-04 at 5 36 38 pm](https://user-images.githubusercontent.com/4188088/199908227-383091a9-429e-4d81-84a8-85141068c039.png)

Does that seem ok to you?

mkllnk commented:
Yes, that's fine for now. It's a pre-existing problem. And I hope that it can be improved when removing AngularJS.

luisramos0 commented on db/migrate/20220919063831_add_opened_notification_at_to_order_cycle.rb:
```diff
@@ -0,0 +1,7 @@
+
+class AddOpenedNotificationAtToOrderCycle < ActiveRecord::Migration[6.1]
+  def change
+    add_column :order_cycles, :opened_notification_at, :timestamp
```
Is this field on the OC meant to mark that a webhook was called? I'd recommend the name of the column to mention that.
I think this should not be on this OC table but rather on a specific webhook calls table/log. This way we are adding complexity to the critical/core OC table without a real need.

dacook replied:
Yes, I think I chose this method to keep it simple for the first iteration and get it released.
But when I pick this up next year I'll consider that, it seems like a much better approach.

dacook replied:
Hmm coming back to this I have a different answer now. We're not intending to log when/if a webhook was called. 
This is a flag to ensure the order cycle "open" state is triggered once, and only once. The application code then schedules any webhooks needing to be delivered. I think  I'll leave this as-is for now.

luisramos0 reviewed (changes requested):
Hi David!
This is a good start! I left a few comments as there are a couple of code/DB design problems imo.

Webhooks usually are named, to have webhooks for multiple events in the system we need named webhooks. For this first PR I'd recommend we have on the webhooks table a "webhook_type" column that we make "order_cycle_opened". On the user page this is a select box with only one option for now: so the user can add a webhook endpoint for a particular event, in this case "Order Cycle Opened".
As mentioned below on not overloading the OC table, I'd create webhook_log table that the Job would fill when the webhook is called, including user_id, oc_id (this can be a generic column like "event_target_object_id" so that other events can put other data in there), etc. This is better than having webhook log data scattered around the data model.

Good luck with this. This is a good start to a great feature!

dacook commented on app/jobs/webhook_delivery_job.rb:
```diff
@@ -0,0 +1,45 @@
+
+# Deliver a webhook payload
+# As a delayed job, it can run asynchronously and handle retries.
+class WebhookDeliveryJob < ApplicationJob
```
The separate job was intended to ensure that delays in delivering webhook(s) didn't impact on other services (ie the SubscriptionPlacementJob).

dacook replied:
For reference:

> Why do you need this to be a job? 

  - delivering webhooks can take time (slowing down the main process), 
  - webhooks can have intermittent errors (that can be resolved by retry) and 
  - there could be multiple happening at once (they could be performed more timely by parallel workers)

> If a job is required I'd still put all the logic in a service WebhookDeliveryService.

Why is this? To allow re-use? Is it bad practice to have app logic in a job? 
I think the logic should stay in the job file, there's not much logic in there.

luisramos0 replied:
I understand, I'd still put all logic in a service.

luisramos0 replied:
Thanks for the explanation!

dacook commented on app/jobs/order_cycle_opened_job.rb:
```diff
@@ -0,0 +1,29 @@
+  def perform
+    ActiveRecord::Base.transaction do
+      recently_opened_order_cycles.find_each do |order_cycle|
+        SubscriptionPlacementJob.perform_later(order_cycle.id)
```
I thought that it would be good to have one place where we define the logic selecting and processing order cycles as "opened".
It does seem strange to be creating so many jobs, yes, but it helps "separate concerns" in other ways I guess. I thought it would be more robust this way.

luisramos0 replied:
I understand but I think this field name is very misleading. What is a notification? I'd use the same terminology everywhere so that code readers will know what this means. It's a webhook. It could be called opened_webhook_sent_at.

I still think it would be much better and very easy to implement this outside the OC table (there's already lots of complexity in OCs).
A table where webhooks calls are registered is important imho. This would be one of them.

dacook replied:
I think I agree, but this would be better considered as part of https://github.com/openfoodfoundation/openfoodnetwork/issues/10457. So for now, I plan to rename this column to  match specifically what it refers to.

dacook commented on app/jobs/application_job.rb:
```diff
@@ -0,0 +1,9 @@
+# frozen_string_literal: true
+
+class ApplicationJob < ActiveJob::Base
+  # Automatically retry jobs that encountered a deadlock
```
This was produced by the generator (rails generate job) when I generated a new job class, and provides potentially useful configuration and documentation. So I thought it worth keeping. 
But another argument I can think of is that it could get out of date (new versions of rails might generate it differently), so it would be better to only include it if needed.
So I agree to remove it.

dacook replied:
It's now been included on another PR so this can be ignored.

dacook commented:
Thank you Luis, this is really helpful feedback! I've responded to some comments and will take this feedback on when I pick the task back up next year.

dacook commented:
User Interface: (some tidying up to be done..)



![Screen Shot 2023-02-07 at 3 29 32 pm](https://user-images.githubusercontent.com/4188088/217148677-61629328-cf69-4055-b1e6-af248bf1514b.png)

![Screen Shot 2023-02-07 at 3 29 43 pm](https://user-images.githubusercontent.com/4188088/217148686-bafe40c6-56bd-4694-8cec-2e43e7bd6096.png)

lin-d-hop commented:
So exciting :grin:

luisramos0 commented on app/jobs/order_cycle_webhook_job.rb:
```diff
@@ -0,0 +1,18 @@
+class OrderCycleWebhookJob < ApplicationJob
+  def perform(order_cycle_id, event)
+    order_cycle = OrderCycle.find(order_cycle_id)
+    data = order_cycle
```
I'd call this variable webhook_payload so it's easier to see what it will be used for.

luisramos0 commented on app/jobs/order_cycle_opened_job.rb:
```diff
@@ -0,0 +1,29 @@
+    ActiveRecord::Base.transaction do
+      recently_opened_order_cycles.find_each do |order_cycle|
+        SubscriptionPlacementJob.perform_later(order_cycle.id)
+        OrderCycleWebhookJob.perform_later(order_cycle.id, 'order_cycle.opened')
```
This type of event 'order_cycle.opened' should be extracted to constant somewhere.

dacook replied:
I'm not sure where/how exactly yet, but will try to address this in my next update next week.

luisramos0 reviewed (commented):
Hi David, I leave a few more suggestions.

dacook commented:
Thanks @luisramos0 , this is valuable feedback. It seems this system design is not quite right, but I'm not clear on the best path and would like to communicate with the team. I will try to process this and perhaps create a forum post to discuss.

dacook commented:
Hi @mkllnk , I'm asking you first because we originally discussed this together and hopefully it makes sense to you. Let me know if we should move to the forum. I've outlined the current structure in the PR description above.
But, as you might see in some of the commit messages, I intend to make some changes thanks to Luis' comments. 

**4. SubscriptionPlacementJob called from OrderCycleOpenedJob**
I moved the existing `SubscriptionPlacementJob` to be triggered from here, along with the new `OrderCycleWebhookJob`. This was to create a consistent way to trigger the event "order cycle opened" within our system. However Luis [remarks](https://github.com/openfoodfoundation/openfoodnetwork/pull/9687#discussion_r1048500985) that it's a bad idea to mix two different tasks. Also I think it's not strictly necessary for this PR (it already worked fine before). 

If it was necessary, I now think that this would be better achieved using an internal events system, like [described here](https://blog.carbonfive.com/evented-rails-decoupling-complex-domains-in-rails-with-domain-events/). That sounds cool and would probably help us scale to the next level..

But do you agree about just leaving `SubscriptionPlacementJob` as it is for now?

**3. OrderCycleWebhookJob: creating and scheduling the payloads**:
Luis suggests this logic should be in a service or elsewhere. I think I agree, and I think performance would be fine if we just called it directly from the `OrderCycleWebhookJob`, because I don't imagine there being many subscribers to an order cycle (currently it's limited to the owner of the coordinator or distributors). 

It could be taken further with a specific type of webhook class. A framework that defines and enforces certain parts of the webhook and payload structure. We'll probably also want to consider versioning and swagger documentation (but [i think that should be in the future](https://github.com/openfoodfoundation/openfoodnetwork/issues/9616#issuecomment-1425133692)).

If we did need to optimise, then I would investigate using an internal event system with async jobs (sidekiq) to solve the problem (also mentioned in the linked article). That's essentially what I built already, but instead of adding yet more specific job classes, an event system would be a more generic and scalable solution.

**5. OrderCycleOpenedJob: marking order cycles as having been 'notified'**
Luis [makes a good point](https://github.com/openfoodfoundation/openfoodnetwork/pull/9687#discussion_r1107782609) that the name `opened_notification_at` doesn't relate to what it's doing. It was intended to say "all tasks related to the opening of an order cycle have been triggered", and prevent accidentally triggering them again.

He recommends having a separate table as a log. I can't think of what to name that yet.

But I'm thinking why not maintain a more generic state field instead (`state: :upcoming > :open > :closed`). This doesn't say that the webhooks were generated, just that the `OrderCycleOpenedJob` ran and doesn't need to run again.
We can then also use this to replace `OrderCycleClosingJob`'s `processed_at`, which really means `state: :closed`.
What do you think of that approach?

dacook commented:
Thanks for the discussion @mkllnk and @luisramos0. Based on the above, I've created a new issue for a path forward with event management: https://github.com/openfoodfoundation/openfoodnetwork/issues/10457

For this PR, I will avoid intersecting with that as much as possible.

mkllnk reviewed (commented):
Wow, great work. Finally a complete version. My comments are just questions and minor improvements. It's great!

dacook commented on Gemfile:
```diff
@@ -185,3 +185,5 @@ group :development do
   gem 'rack-mini-profiler', '< 3.0.0'
 end
+
+gem "http", "~> 5.1"
```
I can't remember! But it did the necessary job.

That's a great reference. Faraday is a clear winner in terms of popularity, and another webpage also shows the highest issue closure rate. It looks like it is super flexible and might be a good choice for HTTP operations in the future. But apparently we haven't needed it thus far. 

I can see in git history we once had HTTParty, which also looks like a solid option. But apparently we don't need it anymore.

So.. 🤷 this works, but I think Faraday would be a safer option, as it should be able to cater for our future needs and we can just learn the one API. It doesn't look like it would take much re-implementation. I'll have a look at the remaining comments before offering an opinion on if we should change or not.

dacook commented on app/jobs/webhook_delivery_job.rb:
```diff
@@ -18,14 +20,24 @@ def perform(url, event, payload)
+    # Note that private_address_check provides some methods for pre-validating,
+    # but they're not as comprehensive and so unnecessary here. Simply
+    # momentarily opening sockets probably can't cause DoS or other damage.
+    PrivateAddressCheck.only_public_connections do
+      response = HTTP
+        .timeout(30)
+        .headers(
+          'User-Agent' => 'openfoodnetwork_webhook/1.0',
+          'Content-Type' => 'application/json',
+        )
+        .post(url, body: body.to_json)
+    end
```
👍 Good suggestion, noted.

dacook reviewed (commented):
Thanks @mkllnk , this helps tidy things up!

I've applied your suggestions as new commits which I'll rebase once you've reviewed.

dacook commented:
Sorry I requested review prematurely.
This is now rebased, and all tests are green.

abdellani commented on db/schema.rb:
```diff
@@ -310,6 +310,7 @@
     t.datetime "processed_at"
     t.boolean "automatic_notifications", default: false
     t.boolean "mails_sent", default: false
+    t.datetime "opened_at"
```
I have one quick question, just to confirm that understand the logic :)
if `opened_at` is `null`, this means that the webhook endpoint is not contacted yet.
Once all the webhook endpoints are contacted, we update this field.
is that correct?

mkllnk replied:
Almost. Once we recognised that the order cycle opened and we dispatched events then `opened_at` is set. So the background jobs to deliver webhooks are enqueued at this point but may not be executed yet. It happens in parallel and the endpoints are contacted at the same time or after the `opened_at` field is set.

abdellani replied:
I understand better now.
Thank you

mkllnk commented:
> I've applied your suggestions as new commits which I'll rebase once you've reviewed.

Oops, I didn't get to review before the rebase but I'll do my best in retrospect.

mkllnk reviewed (approved):
Great work. I just went through my previous comments and found them all addressed. :+1: 

If you introduced new issues then someone else has to spot them. :wink: 

But this is really good, solid work.

dacook commented:
> Oops, I didn't get to review before the rebase but I'll do my best in retrospect.

Sorry, that was a mistake in what I said. I just meant that I rebased _from master_ to resolve conflicts. 
Anyway, thanks for the review, I've now also rebased to incorporate the changes (including the position of the `FailedRequestError` class in "Raise error on server error").

dacook commented:
Now ready for a second review. @luisramos0  I understand if you don't have time, but if you do, would you like to re-review?

rioug commented on app/services/order_cycle_webhook_service.rb:
```diff
@@ -11,6 +11,9 @@ def self.create_webhook_job(order_cycle, event)
     webhook_endpoints = order_cycle.coordinator.owner.webhook_endpoints
 
+    # Plus unique endpoints for distributor owners (ignore duplicates)
+    webhook_endpoints |= order_cycle.distributors.map(&:owner).flat_map(&:webhook_endpoints)
```
TIL about `|=` operator, if I understand correctly it is equivalent to :
```ruby
webhook_endpoints.union(order_cycle.distributors.map(&:owner).flat_map(&:webhook_endpoints))
```
I wonder if we should be using `union` to be more explicit on what's is happening here ?

dacook replied:
Yes, I think that's a good idea, it's pretty obscure. I had to look up the operator before I used it.

dacook replied:
Ah, except that `union` doesn't update the object itself. Also this object is array-like and doesn't support `union` directly, so it would need to be more like:

```ruby
webhook_endpoints = webhook_endpoints.to_a.union(order_cycle.distributors.map(&:owner).flat_map(&:webhook_endpoints))
```

I can think of a few other ways to write this, but I think I'll just stick with the original now...

rioug commented on spec/jobs/order_cycle_opened_job_spec.rb:
```diff
@@ -26,5 +26,37 @@
+
+      # Start two jobs in parallel:
+      threads = [
+        Thread.new { OrderCycleOpenedJob.perform_now },
+        Thread.new { OrderCycleOpenedJob.perform_now },
```
TIL how to test for race condition with thread :+1:

rioug commented on spec/models/order_cycle_spec.rb:
```diff
@@ -551,6 +551,30 @@
+    it "reset opened_at if open date change in future" do
+      expect(oc.opened_at).to_not be_nil
+      oc.update!(orders_open_at: 1.week.from_now, orders_close_at: 2.weeks.from_now)
+      expect(oc.opened_at).to be_nil
```
I don't think we enforce a limit on the number of `expect` per example (maybe it's something to look at), but I like to try to stick to one if possible.
I think it would be better to set the value of `oc.opened_at` instead of checking if it's not nil. What do you think ?

dacook replied:
That's probably a good idea, at least for unit tests. For integration/system specs, I think it's necessary to have multiple expectations.

In this case, I was checking a pre-condition, but the better way to do that would be with `change`. I'll update.

dacook replied:
Updated in new commit.

rioug replied:
> That's probably a good idea, at least for unit tests. For integration/system specs, I think it's necessary to have multiple expectations.
True, I forgot about system specs

rioug commented on spec/models/order_cycle_spec.rb:
```diff
@@ -551,6 +551,30 @@
+    it "it does not reset opened_at if open date is changed to be earlier" do
+      expect(oc.opened_at).to_not be_nil
+      oc.update!(orders_open_at: 3.days.ago)
+      expect(oc.opened_at).to_not be_nil
```
Same as above

rioug commented on spec/models/order_cycle_spec.rb:
```diff
@@ -551,6 +551,30 @@
+    it "it does not reset opened_at if open date does not change" do
+      expect(oc.opened_at).to_not be_nil
+      oc.update!(orders_close_at: 1.day.from_now)
+      expect(oc.opened_at).to_not be_nil
```
Same as above

rioug commented on spec/controllers/webhook_endpoints_controller_spec.rb:
As Filipe mentioned a little while ago,  Rspec recommend moving away from controller spec, and use request spec instead. I am wondering if we should follow that recommendation, or just use system spec instead ?

dacook replied:
Oh yes, I forgot that. My general opinion is to have small, efficient unit tests to cover all possible edge cases, then to cover the basics with an integration (system) test to make sure everything works together. But I haven't thought that much about it.

Looking at the [recommendation](https://medium.com/just-tech/rspec-controller-or-request-specs-d93ef563ef11), I can see how request specs are more comprehensive, and performant, so I agree about using them over controller specs in principle. But I think the team needs to agree on an approach.

rioug reviewed (approved):
Looks good :clap: I left a few comments but nothing that should block the PR.
I like your use of `pending` specs for work to be done, which I should start using myself

dacook commented:
Ready to test!

filipefurtad0 commented:
Hey @dacook,

Thanks, really, for all the detail in the description of the PR. It comes really handy.

1) **Receiving payload, after OC is opened** :heavy_check_mark: 

After enabling the API key for a user which is coordinator and distributor of an OC, we can check the UI:

![image](https://user-images.githubusercontent.com/49817236/225689597-6c7ce133-add5-43fc-9e25-429899978875.png)

Creating the webhook displays:

![image](https://user-images.githubusercontent.com/49817236/225689727-17f3f2db-2e01-4239-9de4-732bde0a4a36.png)

Proceeded to create an order cycle to open in a future time, taking note of the order cycle id (`"id": 3671`) I've found it easier to use the procedure to open the OC as [per this wiki](https://github.com/openfoodfoundation/openfoodnetwork/wiki/Testing-Subscriptions) -> thanks again to @luisramos0 for introducing this nice testing code, back in the day!! :heart: This way, we don't need to consider timezone differences when creating the OC, and there is no wait-time till it opens :tada:

Some moments afterwards, the webhook site was updated :rocket:  we can check the payload below: 

![image](https://user-images.githubusercontent.com/49817236/225693483-b72156a1-0a51-4aa8-8e3d-7caea9f3e281.png)

And compare it to the info we get from the rails console, on the server:

```

[#<OrderCycle:0x00000000090d9540
  id: 3671,
  name: "webhook_OC",
  orders_open_at: Thu, 16 Mar 2023 12:32:55.095320000 EDT -04:00,
  orders_close_at: Thu, 16 Mar 2023 13:02:55.095421000 EDT -04:00,
  coordinator_id: 1822,
  created_at: Thu, 16 Mar 2023 12:38:36.100790000 EDT -04:00,
  updated_at: Thu, 16 Mar 2023 12:49:23.926191000 EDT -04:00,
  processed_at: nil,
  automatic_notifications: false,
  mails_sent: false,
  opened_at: Thu, 16 Mar 2023 12:49:23.926191000 EDT -04:00>]
```

All good here!

2) **Not receiving any payload, after deleting the webhook endpoint**  :heavy_check_mark: 

After deleting the endpoint, we receive no payload, when the OC is re-opened.

3) **Re-entering the webhook endpoint on ** `/account#/developer_settings`  :heavy_check_mark: 
...works as before, and delivers the payload as expected:

![image](https://user-images.githubusercontent.com/49817236/225699777-4b96e83f-f94f-40bd-b790-f32bbcf8f502.png)

4) **Opening an order cycle in which the user is _supplier and distributor_**  :heavy_check_mark: 

Also works, payload is delivered with the respective data (`"id": 3661`):

![image](https://user-images.githubusercontent.com/49817236/225700618-9445fb76-4442-49c9-a4fa-9859dea557a0.png)

The UI is according to specification, functionality as well, data is consistent.
I'd say this looks great - awesome! :clap: :clap: :clap: 
Merging.

dacook commented:
## 📚 Draft documentation
See here: https://github.com/openfoodfoundation/ofn-UserGuide/issues/41#issuecomment-1473118352
